app/services/sso/openid_connect.rb

```diff
@@ -20,6 +22,12 @@ def authenticated?
       self.user = User.current.presence || authenticate!
     end
 
+    def logout_url
+      default_logout_url = Setting['login_delegation_logout_url']
+      return unless mod_auth_openidc_session_exist?
+      default_logout_url + "/redirect_uri?logout=" + CGI.escape(default_logout_url)
```

Comment:
I still don't understand the `redirect_uri` part. When I read https://github.com/zmartzone/mod_auth_openidc/wiki#9-how-do-i-logout-users it states:
> You can kill the session from your application by redirecting the user to the OIDCRedirectURI with a parameter named logout.

To me that reads like we should tell the user to literally copy `OIDCRedirectURI` to `login_delegation_logout_url` and all we need to do is add a `logout` parameter.

I think you're also sending the user in a redirect loop now. Why can't you use `controller.extlogout_users_path` as `logout` as we do in `app/services/sso/apache.rb`?

Reply:
@ekohl its the same thing. The OIDCRedirectURI is equal to `https://foreman.com/users/extlogin/redirect_uri` and we need to add a logout parameter to it, which is same as `default_logout_url + "/redirect_uri?logout="`

Reply:
So disagree with that. Some testing:
```console
# keycloak-httpd-client-install --app-name foreman-openidc --keycloak-server-url $KEYCLOAK_URL --keycloak-admin-username $KEYCLOAK_USER --keycloak-realm $KEYCLOAK_REALM --keycloak-admin-realm master --keycloak-auth-role root-admin -t openidc -l /users/extlogin --oidc-redirect-uri /users/extlogin/logout
enter admin password: 
[Step  1] Assure HTTP config directory is present
[Step  2] Assure HTTP federation directory is present
[Step  3] Set up template environment
[Step  4] Build OIDC httpd config file
[Step  5] Build Keycloak OIDC clientRepresentation
[Step  6] Connect to Keycloak Server as admin
CommunicationError: Unable to open connection to "https://keycloak.wisse.example.com" as admin: ('Connection aborted.', gaierror(-2, 'Name or service not known'))
# cat /etc/httpd/conf.d/foreman-openidc_oidc_keycloak_ssl-realm.conf
OIDCClientID 84398834241a-foreman-openidc
OIDCProviderMetadataURL https://keycloak.wisse.example.com/auth/realms/ssl-realm/.well-known/openid-configuration
OIDCCryptoPassphrase af46dcee623d
OIDCClientSecret b6814fbf722c
OIDCRedirectURI https://84398834241a/users/extlogin/logout
OIDCRemoteUserClaim sub

<Location /users/extlogin>
    AuthType openid-connect
    Require valid-user
</Location>
```

It just happens to be that `redirect_uri` is the default so hardcoding it here looks incorrect.

Reply:
So based on an offline discussion, I think we should have some thing like (untested):

```ruby
def logout_url
  redirect_uri = Setting['login_delegation_logout_url'] || "#{controller.main_app.extlogin_users_path}/redirect_uri"
  "#{redirect_uri}?logout=#{CGI.escape(controller.extlogout_users_path)}"
end

This allows users to override the `OIDCRedirectURI` and set it via `login_delegation_logout_url` but we have defaults that match what `keycloak-httpd-client-install` does. It also redirects users to the extlogout page as @ares showed. It allows `mod_auth_oidc` to intercept and destroy the session and shows the user a page where they're logged out.